spec/controllers/oauth/authorizations_controller_spec.rb

```diff
@@ -61,7 +78,110 @@
+
+          expect(code_challenge_input).to be_present
+          expect(code_challenge_method_input).to be_present
+
+          expect(code_challenge_input.attr('value')).to be_nil
+          expect(code_challenge_method_input.attr('value')).to be_nil
+        end
+
+        include_examples 'shows authorize and deny buttons'
+      end
+
+      context 'when using PKCE' do
+        subject do
+          get :new, params: { client_id: app.uid, response_type: 'code', redirect_uri: 'http://localhost/', scope: 'read', code_challenge_method: pkce_code_challenge_method, code_challenge: pkce_code_challenge }
+        end
+
+        let(:pkce_code_challenge) { SecureRandom.hex(32) }
+        let(:pkce_code_challenge_method) { 'S256' }
+
+        context 'when using S256 code challenge method' do
+          it 'returns http success' do
+            subject
+            expect(response).to have_http_status(200)
+          end
+
+          it 'includes the PKCE values in the response' do
+            subject
+
+            code_challenge_input = body.at_css('.oauth-prompt input[name=code_challenge]')
+            code_challenge_method_input = body.at_css('.oauth-prompt input[name=code_challenge_method]')
+
+            expect(code_challenge_input).to be_present
+            expect(code_challenge_method_input).to be_present
+
+            expect(code_challenge_input.attr('value')).to eq pkce_code_challenge
+            expect(code_challenge_method_input.attr('value')).to eq pkce_code_challenge_method
+          end
+
+          include_examples 'shows authorize and deny buttons'
+        end
+
+        context 'when using plain code challenge method' do
+          subject do
+            get :new, params: { client_id: app.uid, response_type: 'code', redirect_uri: 'http://localhost/', scope: 'read', code_challenge_method: pkce_code_challenge_method, code_challenge: pkce_code_challenge }
+          end
+
+          let(:pkce_code_challenge_method) { 'plain' }
+
+          it 'returns http success' do
+            subject
+            expect(response).to have_http_status(400)
+          end
+
+          it 'does not include the PKCE values in the response' do
+            subject
+
+            code_challenge_input = body.at_css('.oauth-prompt input[name=code_challenge]')
+            code_challenge_method_input = body.at_css('.oauth-prompt input[name=code_challenge_method]')
+
+            expect(code_challenge_input).to_not be_present
+            expect(code_challenge_method_input).to_not be_present
+          end
+
+          it 'does not include the authorize button' do
+            subject
+
+            authorize_button = body.at_css('.oauth-prompt button[type="submit"]')
+
+            expect(authorize_button).to_not be_present
+          end
+
+          it 'includes an error message' do
+            subject
+            expect(error_message).to match I18n.t('doorkeeper.errors.messages.invalid_code_challenge_method')
+          end
```

Comment:
All those new examples here seem to test the user interface and as such do not belong in a "controller spec".

There are two alternative options:

* If you just want to make sure certain things are present in the html template, then a "view spec" would be the correct place. Additionally you _could_ add a test here that makes sure the controller actually renders that template.
* If you want to make sure that a certain kind of client request leads to the UI being in a certain state, this should be a "system spec". If you go down that route, you might want to have these be "driven_by" `:rack_test` which is a lot faster than spinning up a browser for this (see https://rspec.info/features/6-0/rspec-rails/system-specs/system-specs/)

Reply:
I do agree, though as we were already doing the controller spec like this (assertions on response body) I'd like to run with this as-is, and create a follow up issue to address controller vs request vs system tests for OAuth flows

Reply:
I feel like system specs are probably most relevant for OAuth flows, not controller or request or view specs

Maybe we'd have a controller spec that just asserts out custom views for doorkeeper get rendered?

(also, I try really hard to limit scope of changes due to frequently being asked to break PRs down into smaller chunks, which takes significant time)

Reply:
Just to be super-clear: I am not suggesting creating generic oauth flow system specs. Just moving these here out to either a view or a (single) system spec file. This would not widen the scope of this PR, just make better use of the different kinds of spec types `rspec-rails` provides.

Reply:
I'm going to wait for @ClearlyClaire's input here, since I don't want to risk overcomplicating this PR by rewriting a known working test too heavily (as it is I only made the changes I did because the assertions were quiet weak & could lead to tests passing when they shouldn't

Reply:
I agree with @oneiros that this kind of tests belong in system tests. I'd suggest looking at the existing `spec/system/oauth_spec.rb` tests to see if those could be moved there instead without a lot more effort, but if that amount of work is prohibitive at the moment, I'm ok with the consolidation happening in another PR (note that I have not actually reviewed the specs yet, just quickly read through them, and I'm commenting on the general approach)

Reply:
Yeah, I'd like to do a review of the oauth specs as a separate task after implement PKCE. Only reason I noticed was because they were already doing stuff like

```
        expect(response.body).to match(/Authorize/)
```

Which isn't great at all, because the string Authorize could appear *anywhere* on the response body.

Reply:
Have created https://github.com/mastodon/mastodon/issues/31150 to track the OAuth Specs state.